### ISSUE TITLE
Search Function

### ISSUE THREAD

**Issue by thombruce** [thombruce/journal#3]
We already save encrypted text from Tiptap given by the value of document.textContent. ~~Note: This value is poor, as it ignores non-space whitespace, resulting in strings like: "Paragraph oneParagraph two". It would be better to use the DOM's innerText property. Should look into how capable Tiptap (or underlaying Prosemirror) might be of providing that.~~

The text we save is encrypted. It should not be unencrypted anywhere but on the user's machine. Do not save plain text.

This means we need to fetch each node individually, decrypt it, and then search it. Actually... see footnote.

As a first effort towards searchable notes, we'll add a comparison to the `.map()` function in the gun chain. This is... "the easy way", but slow if we ever have to deal with many records. The approach in our footnote below might be more performant at scale, while still maintaining privacy.

---

Footnote:

It may be possible to individually encrypt search keywords for search against an encrypted word bank. Let's not go straight for this option, but it could be worth looking into... Essentially, both the word bank's words and the query words would be encrypted using the user's credentials, so that we can guarantee matching, even without knowing the words.

**Comment by thombruce** [comment]
Note, the poor quality of text retrieval from Tiptap has now been resolved.

`state.doc.textContent` was replaced with `state.doc.textBetween(0, state.doc.content.size, ' ')` in store/editor/actions.js.

These functions are documented in prosemirror-model starting at line 91 here: https://github.com/ProseMirror/prosemirror-model/blob/master/src/node.js#L91

```js
  // :: string
  // Concatenates all the text nodes found in this fragment and its
  // children.
  get textContent() { return this.textBetween(0, this.content.size, "") }

  // :: (number, number, ?string, ?string) → string
  // Get all text between positions `from` and `to`. When
  // `blockSeparator` is given, it will be inserted whenever a new
  // block node is started. When `leafText` is given, it'll be
  // inserted for every non-text leaf node encountered.
  textBetween(from, to, blockSeparator, leafText) {
    return this.content.textBetween(from, to, blockSeparator, leafText)
  }
```

**Comment by thombruce** [comment]
Fully implemented as of 7ed22f554e4f2aca13592ca4651e2b91eb7786f0
